### ISSUE TITLE
Flowdroid 2.9 compilation error "Cannot find symbol xxx"

### ISSUE THREAD

**Issue by occia** [secure-software-engineering/FlowDroid#405]
Please examine each of the following points so that we can help you as soon and best as possible.

**Describe the bug**
I followed the building steps to build flowdroid 2.9. But still meets the compilation error "cannot find symbol xxx". I also tried build flowdroid from developer branch but also meets the same error.

**My Build Steps**

```bash
# 1. download the tar
wget  https://github.com/secure-software-engineering/FlowDroid/archive/refs/tags/v2.9.tar.gz

# 2. build
mvn -DskipTests install
```

**Error Information**
```bash
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/cfg/LibraryClassPatcher.java:[18,25] cannot find symbol
  symbol:   class LocalGenerator
  location: package soot.javaToJimple
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/entryPointCreators/BaseEntryPointCreator.java:[48,25] cannot find symbol
  symbol:   class LocalGenerator
  location: package soot.javaToJimple
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/entryPointCreators/BaseEntryPointCreator.java:[102,19] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/codeOptimization/InterproceduralConstantValuePropagator.java:[39,25] cannot find symbol
  symbol:   class LocalGenerator
  location: package soot.javaToJimple
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/cfg/LibraryClassPatcher.java:[231,17] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.cfg.LibraryClassPatcher
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/cfg/LibraryClassPatcher.java:[231,41] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.cfg.LibraryClassPatcher
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/entryPointCreators/BaseEntryPointCreator.java:[146,33] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/entryPointCreators/BaseEntryPointCreator.java:[226,17] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/entryPointCreators/BaseEntryPointCreator.java:[226,41] cannot find symbol
  symbol:   class LocalGenerator
  location: class soot.jimple.infoflow.entryPointCreators.BaseEntryPointCreator
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/codeOptimization/InterproceduralConstantValuePropagator.java:[511,57] cannot find symbol
  symbol: class LocalGenerator
[ERROR] /data/xxx/workspace/FlowDroid-2.9/soot-infoflow/src/soot/jimple/infoflow/codeOptimization/InterproceduralConstantValuePropagator.java:[511,88] cannot find symbol
  symbol: class LocalGenerator
```

My java version is openjdk 11 and  mvn version is 3.6.0.
Do you have any idea for fixing this?
Thank you.

**Comment by occia** [comment]
After some investigation I solved this issue. The compilation error is caused by the wrong version of soot specified in `FlowDroid/soot-infoflow/pom.xml`. 

# compilation error detail
In that xml, the soot version is specified as `4.3.0-SNAPSHOT`, which is a nightly build version that updates itself every time when the soot repo has a new commit. The flowdroid 2.9 is released in 2021 Jul and some apis of soot have been changed during the past 4 months. Therefore, you saw sth like `soot.javaToJimple.LocalGenerator cannot be found`.

# my fix
My idea for the fix is to build a local soot jar which has a compatible version with flowdroid 2.9 and build flowdroid using that local version soot jar. I choose the commit  9f34d1fcc218596e8efb81742c9f160100c896d7 for soot since it is the latest commit before the release date of flowdroid 2.9.
 
```bash
# build suitable soot version
git clone git@github.com:soot-oss/soot.git
cd soot
git checkout 9f34d1fcc218596e8efb81742c9f160100c896d7
mvn clean compile assembly:single

# build flowdroid 2.9 using that soot
wget https://github.com/secure-software-engineering/FlowDroid/archive/refs/tags/v2.9.tar.gz
# unzip the v2.9 ...
# follow this link (https://stackoverflow.com/a/28762617/5598892) to modify the soot-infoflow/pom.xml and install soot into local-mvn-repo
# after that, build flowdroid
mvn -DskipTests install
```

# misc
The method is also available for flowdroid 2.8 if you need that version, just to pick a suitable commit of soot for flowdroid.

@StevenArzt  I think this repo needs some documentation about this issue since soot didn't provide the way to download old jars that is suitable for flowdroid 2.8/2.9. Or do you have any better idea for fixing this build issue? If not and if you need the documentation, I'm happy to create a PR.

**Comment by StevenArzt** [comment]
The easiest solution is to use the code from the ``develop`` branch, which is the current development version and works fine with the current Soot snapshots. We should definitely create a new official FlowDroid release some time soon.
